### PR TITLE
Add long polling to update unread counts in navbar.

### PR DESCRIPTION
**Pull request by dessalines**
Currently have this set to 30 seconds, either that or 1m seems good.

### CODE REVIEW

File: src/shared/utils.ts [hunk 1]
```diff
@@ -1490,3 +1491,14 @@ export function newVote(voteType: VoteType, myVote?: number): number {
...
+
+function sleep<T>(millis: number): Promise<T> {
+  return new Promise(resolve => setTimeout(resolve, millis));
+}
+
+/**
+ * Polls / repeatedly runs a promise, every X milliseconds
+ */
+export function poll(promiseFn: any, millis: number) {
+  promiseFn().then(sleep(millis).then(() => poll(promiseFn, millis)));
```
Comment (SleeplessOne1917) [inline src/shared/utils.ts:1503]: I think this would be easier to read with awaits.

**Comment by SleeplessOne1917** [comment]
I saw unread messages, reports, and applications getting polling. Is there one needed for replies as well?

File: src/shared/utils.ts (hunk 1 above)
Comment (dessalines) [reply src/shared/utils.ts]: K fixed.

**Comment by dessalines** [comment]
unreadInboxCount includes unread replies, messages, and mentions.

**Comment by sunaurus** [comment]
Sorry, I realized that I was commenting on #1148 when I meant to comment here 🤦 moving my comment:

Many users will have tens (in some cases even hundreds) of tabs open, so it could be extremely useful to either A) completely pause polling or B) just skip invoking the promiseFn in your poll function whenever document.visibilityState != "visible"

For example with the 30 second polling:

If an instance has an average of 1000 users who leave 100 tabs open in the background, then not pausing background polling will generate 288 000 000 extra unnecessary http requests in a day (1000 users * 100 tabs * 2880 requests in a tab per 24h)

**Comment by SleeplessOne1917** [comment]
@sunaurus I recall you linked a browser API that can be used to avoid this problem in the issue comments.

**Comment by sunaurus** [comment]
Yep, it's here: https://developer.mozilla.org/en-US/docs/Web/API/Document/visibilityState

File: src/shared/utils.ts [hunk 2]
```diff
@@ -1490,3 +1491,16 @@ export function newVote(voteType: VoteType, myVote?: number): number {
...
+
+function sleep<T>(millis: number): Promise<T> {
+  return new Promise(resolve => setTimeout(resolve, millis));
+}
+
+/**
+ * Polls / repeatedly runs a promise, every X milliseconds
+ */
+export async function poll(promiseFn: any, millis: number) {
+  await promiseFn();
```
Comment (sunaurus) [inline src/shared/utils.ts:1503]: In theory, this should be enough, but I'm not able to test it myself at the moment:

```suggestion
  if (window.document.visibilityState !== "hidden") {
    await promiseFn();
  }
```

Comment (dessalines) [reply src/shared/utils.ts:1503]: Ohh I see so this prevents it if you're looking at other tabs?

Comment (sunaurus) [reply src/shared/utils.ts:1503]: Yup, exactly

Comment (dessalines) [reply src/shared/utils.ts:1503]: K sweet, I'll test that shortly.

File: src/shared/utils.ts [hunk 3]
```diff
@@ -1490,3 +1491,16 @@ export function newVote(voteType: VoteType, myVote?: number): number {
     return myVote == -1 ? 0 : -1;
   }
 }
+
+function sleep<T>(millis: number): Promise<T> {
```
Comment (SleeplessOne1917) [inline src/shared/utils.ts:1495]: The generic isn't needed here. You could get away just fine returning `Promise<void>`.

Comment (dessalines) [reply src/shared/utils.ts:1495]: Sry, getting to this now.

**Review by SleeplessOne1917 — APPROVED** [review]

**Pull request merged**